### PR TITLE
Got rid of recursive restart mechanism in `AsyncEffectRenderer`

### PR DESCRIPTION
**Pull request by Lehonti**
This mechanism was very awkward. Now restarts are handled externally in the `LivePreviewManager`.

The `Task` mechanism is finally being used and we are on a good track to implement the async rendering mechanism cleanly!

### CODE REVIEW

File: Pinta.Core/Classes/AsyncEffectRenderer.cs [hunk 1]
```diff
@@ -117,11 +120,20 @@ internal double Progress {
...
 	}
 
+	private readonly object completion_swap_lock = new ();
 	internal void Start (
 		BaseEffect effect,
 		Cairo.ImageSurface source,
 		Cairo.ImageSurface dest)
 	{
+		TaskCompletionSource newCompletionSource = new ();
+		lock (completion_swap_lock) {
```
Comment (cameronwhite) [inline Pinta.Core/Classes/AsyncEffectRenderer.cs:130]: Could you explain why the lock is needed here?

Not sure if I'm missing something, but if this is only called with `IsRendering == false` then it seems like there shouldn't be concurrent access to `completion_source` here?

File: Pinta.Core/Managers/LivePreviewManager.cs [hunk 2]
```diff
@@ -238,10 +239,16 @@ void CleanUp ()
 			chrome.MainWindowBusy = false;
 		}
 
-		void EffectData_PropertyChanged (object? sender, PropertyChangedEventArgs e)
+		async void EffectData_PropertyChanged (object? sender, PropertyChangedEventArgs e)
 		{
-			// TODO: calculate bounds
-			renderer.Start (effect, layer.Surface, LivePreviewSurface);
+			await restartSemaphore.WaitAsync ();
```
Comment (cameronwhite) [inline Pinta.Core/Managers/LivePreviewManager.cs:244]: Similar to the other question, why is a semaphore now required here?

**Review by cameronwhite — COMMENTED** [review]
Nice, moving towards using tasks seems like a cleaner model for restarting / cancelling the renders 👍

**Comment by Lehonti** [comment]
@cameronwhite, responding to your reviews:

Sorry for the mess. It is meant to be cleaned up.

As for the reasoning: at first sight it might seem that there is not concurrent access. However, for some reason, property change events are fired in rapid succession so all of the handlers scramble to `await` the cancellation and then start rendering at the same time, which causes the exception to be thrown. It happens both on my Windows machine and on my Ubuntu virtual machine. I am sure that if you try it it will happen as well. I have prepared commit [3dc5155](https://github.com/PintaProject/Pinta/pull/1109/commits/3dc5155dc5d0e59a1577bee411e6b898d47800e4) if you'd like to see it.

**Comment by cameronwhite** [comment]
Thanks, I could reproduce it too and agree that it happens if there are multiple waiters on the same cancellation task which then start rendering one after the other.
The semaphore does make that work, although there are some wasteful restarts being done - if there are a bunch of rapid changes or the cancellation takes a bit of time, ideally only one render could be started?

**Comment by Lehonti** [comment]
@cameronwhite as a quick solution, what about a counter that tells us how many awaiters are in queue, so that we can skip wasteful restarts if there are? I just implemented this in the last commit

**Comment by cameronwhite** [comment]
The counter idea sounds reasonable to me

- I'm pretty sure it shouldn't need an atomic increment, since this async code should all be executed on the main thread (due to https://github.com/gircore/gir.core/blob/main/src/Libs/GLib-2.0/Internal/MainLoopSynchronizationContext.cs)
- With this approach, is the semaphore still necessary? If the counter instead happens around waiting for the cancellation then I think that might make it unnecessary

**Comment by Lehonti** [comment]
@cameronwhite, if all the coordination happens in the same thread then I agree it would make it unnecessary. I removed the semaphore and made non-atomic increments and decrements, and it works well on my machine.

**Comment by cameronwhite** [comment]
Thanks! Looks good and behaves well in my testing too

**Pull request merged**